File: extension/src/lib/schedule.ts
// Pure scheduling logic. No Chrome APIs here so it can be reasoned about and tested.

import { intervalMinutes, type Settings } from "./settings.ts";

const MIN = 60_000;

export function parseTime(hhmm: string): { h: number; m: number } {
  const [h, m] = hhmm.split(":").map(Number);
  return { h, m };
}

type Quiet = Pick<Settings, "quietStart" | "quietEnd"> & { quietEnabled?: boolean };

export function quietHoursEnabled(s: Quiet): boolean {
  return s.quietEnabled !== false && s.quietStart !== s.quietEnd;
}

/** Minutes since local midnight for a Date. */
function minutesOfDay(d: Date): number {
  return d.getHours() * 60 + d.getMinutes();
}

function toMinutes(hhmm: string): number {
  const { h, m } = parseTime(hhmm);
  return h * 60 + m;
}

/** True when `at` falls inside quiet hours (handles ranges that cross midnight). */
export function isQuiet(at: Date, s: Quiet): boolean {
  if (!quietHoursEnabled(s)) return false;
  const now = minutesOfDay(at);
  const start = toMinutes(s.quietStart);
  const end = toMinutes(s.quietEnd);
  if (start < end) return now >= start && now < end;
  // crosses midnight, e.g. 22:00 → 07:00
  return now >= start || now < end;
}

/** The next moment quiet hours end, at or after `at`. Assumes isQuiet(at) is true. */
export function quietEndAfter(at: Date, s: Quiet): Date {
  const { h, m } = parseTime(s.quietEnd);
  const end = new Date(at);
  end.setHours(h, m, 0, 0);
  if (end.getTime() <= at.getTime()) end.setDate(end.getDate() + 1);
  return end;
}

/**
 * Compute when the next reminder should fire.
 * - base: now + interval
 * - randomize: the interval is stretched or shortened by up to 25 %, so a
 *   1-hour reminder lands somewhere between 45 and 75 minutes from now.
 * - quiet hours: a reminder that would land inside them moves to just after
 *   they end (plus a few random minutes when randomize is on).
 */
export function computeNextFire(now: Date, s: Settings, rand: () => number = Math.random): Date {
  const interval = intervalMinutes(s) * MIN;
  let delay = interval;
  if (s.randomize) {
    const factor = 0.75 + rand() * 0.5; // 0.75 … 1.25
    delay = Math.round(interval * factor);
  }
  let next = new Date(now.getTime() + delay);
  if (isQuiet(next, s)) {
    next = quietEndAfter(next, s);
    if (s.randomize) next = new Date(next.getTime() + Math.round(rand() * 15 * MIN));
  }
  // never sooner than one minute from now (Chrome's floor for packed extensions)
  if (next.getTime() - now.getTime() < MIN) next = new Date(now.getTime() + MIN);
  return next;
}

export function formatInterval(s: Settings): string {
  const m = intervalMinutes(s);
  if (m % 60 === 0) {
    const h = m / 60;
    return h === 1 ? "1 hour" : `${h} hours`;
  }
  if (m > 60) {
    const h = Math.floor(m / 60);
    const rest = m % 60;
    return `${h} h ${rest} min`;
  }
  return `${m} minutes`;
}

export function formatClock(d: Date): string {
  return d.toLocaleTimeString([], { hour: "numeric", minute: "2-digit" }).toLowerCase();
}

export function formatHHMM(hhmm: string): string {
  const { h, m } = parseTime(hhmm);
  const d = new Date();
  d.setHours(h, m, 0, 0);
  return formatClock(d);
}

File: apple/Shared/Schedule.swift
// Pure scheduling logic, a port of extension/src/lib/schedule.ts and
// selection.ts. No system APIs here, so ScheduleTests can check it.

import Foundation

enum TimeOfDay {
    static func isTime(_ s: String) -> Bool {
        s.range(of: #"^([01]\d|2[0-3]):[0-5]\d$"#, options: .regularExpression) != nil
    }

    /// "HH:MM" → minutes since midnight.
    static func minutes(_ hhmm: String) -> Int {
        let parts = hhmm.split(separator: ":").compactMap { Int($0) }
        guard parts.count == 2 else { return 0 }
        return parts[0] * 60 + parts[1]
    }

    /// "8:30 pm", in the person's own locale.
    static func format(_ date: Date) -> String {
        date.formatted(date: .omitted, time: .shortened).lowercased()
    }
}

enum Schedule {
    static let minute: TimeInterval = 60

    static func quietHoursEnabled(_ s: Settings) -> Bool {
        s.quietEnabled && s.quietStart != s.quietEnd
    }

    /// True when `at` falls inside quiet hours (ranges may cross midnight).
    static func isQuiet(_ at: Date, _ s: Settings, calendar: Calendar = .current) -> Bool {
        guard quietHoursEnabled(s) else { return false }
        let c = calendar.dateComponents([.hour, .minute], from: at)
        let now = (c.hour ?? 0) * 60 + (c.minute ?? 0)
        let start = TimeOfDay.minutes(s.quietStart)
        let end = TimeOfDay.minutes(s.quietEnd)
        if start < end { return now >= start && now < end }
        return now >= start || now < end // e.g. 22:00 → 07:00
    }

    /// The next moment quiet hours end, after `at`.
    static func quietEnd(after at: Date, _ s: Settings, calendar: Calendar = .current) -> Date {
        let m = TimeOfDay.minutes(s.quietEnd)
        var end = calendar.date(bySettingHour: m / 60, minute: m % 60, second: 0, of: at) ?? at
        if end <= at { end = calendar.date(byAdding: .day, value: 1, to: end) ?? end }
        return end
    }

    /// When the reminder after `from` should arrive.
    /// - randomize stretches or shortens the interval by up to 25 %;
    /// - a reminder that would land in quiet hours moves to just after they
    ///   end (plus a few random minutes when randomize is on);
    /// - never sooner than one minute after `from`.
    static func nextFire(after from: Date, _ s: Settings, rand: () -> Double = { Double.random(in: 0..<1) },
                         calendar: Calendar = .current) -> Date {
        let interval = TimeInterval(s.intervalMinutes) * minute
        var delay = interval
        if s.randomize { delay = (interval * (0.75 + rand() * 0.5)).rounded() }
        var next = from.addingTimeInterval(delay)
        if isQuiet(next, s, calendar: calendar) {
            next = quietEnd(after: next, s, calendar: calendar)
            if s.randomize { next = next.addingTimeInterval((rand() * 15 * minute).rounded()) }
        }
        if next.timeIntervalSince(from) < minute { next = from.addingTimeInterval(minute) }
        return next
    }

    static func describe(_ s: Settings) -> String {
        let m = s.intervalMinutes
        if m % 60 == 0 { return m == 60 ? "1 hour" : "\(m / 60) hours" }
        if m > 60 { return "\(m / 60) h \(m % 60) min" }
        return "\(m) minutes"
    }
}

/// Which reminder comes next. A cycle is a shuffled list of every reminder;
/// take from the front until it is empty, then reshuffle. The same reminder
/// never shows twice in a row, even across the boundary between cycles.
enum Selection {
    static func pickNext(cycle: [String], last: String?, ids: [String],
                         rand: () -> Double = { Double.random(in: 0..<1) }) -> (id: String, cycle: [String])? {
        var cycle = cycle.filter { ids.contains($0) }
        if cycle.isEmpty {
            cycle = shuffle(ids, rand: rand)
            if cycle.count > 1, cycle[0] == last {
                let other = 1 + Int(rand() * Double(cycle.count - 1))
                cycle.swapAt(0, min(other, cycle.count - 1))
            }
        }
        guard let first = cycle.first else { return nil }
        return (first, Array(cycle.dropFirst()))
    }

    static func shuffle<T>(_ items: [T], rand: () -> Double) -> [T] {
        var a = items
        guard a.count > 1 else { return a }
        for i in stride(from: a.count - 1, to: 0, by: -1) {
            let j = min(i, Int(rand() * Double(i + 1)))
            a.swapAt(i, j)
        }
        return a
    }
}
